scripts/dts/python-devicetree/src/devicetree/sysdtlib.py

```diff
@@ -0,0 +1,640 @@
+            if prop in node.props:
+                _err(f"domain '{node.path}' contains unsupported "
+                     f'property {prop}')
+
+    def _decoded_address_map(self) -> list[AddressMapEntry]:
+        # Decode the domain's 'address-map' property.
+
+        dt = self.node.dt
+        root = dt.get_node('/')
+        if '#address-cells' not in root.props:
+            _err('the root node must have a #address-cells property set')
+        cluster = self.cpu_cluster
+
+        # Number of bytes in each node-address field.
+        node_address_size = 4 * cluster.ranges_address_cells
+        # Number of bytes in each root-node-address field.
+        root_node_address_size = 4 * root.props['#address-cells'].to_num()
+        # Number of bytes in each length field.
+        length_size = 4 * cluster.ranges_size_cells
+
+        # Split the address map property into quartets.
+        entry_size = (node_address_size +
+                      4 +  # ref-node is a phandle
+                      root_node_address_size +
+                      length_size)
+        entries = _slice(self.cpu_cluster.node,
+                        self._address_map_prop,
+                        entry_size,
+                        f'''\
+4 * #ranges-address-cells (= {node_address_size}) + \
+4 (ref-node phandle) + \
+4 * root #address-cells (= {root_node_address_size}) + \
+4 * #ranges-size-cells (= {length_size})''')
+
+        # Enumerate the quartets
+        ref_node_offset = node_address_size
+        root_node_address_offset = ref_node_offset + 4
+        length_offset = root_node_address_offset + root_node_address_size
+        ret = []
+        for entry in entries:
+            ret.append(AddressMapEntry(
+                node_address=to_num(entry[:node_address_size]),
+                ref_node=dt.phandle2node[to_num(entry[ref_node_offset:
+                                                      ref_node_offset + 4])],
+                root_node_address=to_num(entry[root_node_address_offset:
+                                               root_node_address_offset +
+                                               root_node_address_size]),
+                length=to_num(entry[length_offset:
+                                    length_offset + length_size])))
+
+
+        return ret
+
+    @property
+    def _address_map_prop(self):
+        # Get the name of the address map property to use for the
+        # given domain, handling Arm v8-M designs as a special case
+        # that may have 'address-map-secure' properties.
+
+        first_cpu_compat = set(_compatible(self.cpus[0]))
+        if not all(set(_compatible(cpu)) == first_cpu_compat
+                   for cpu in self.cpus):
+            _LOG.warning(f"domain '{self.node.path}' CPUs have differing "
+                         'compatible properties; unexpected address map '
+                         'behavior may result')
+
+        def arm_v8m_address_map():
```

Comment:
Do we envisage similar requirements for other architectures that would need special handling here? I don't know if arc/mips/ppc/risc/xtensa have or would implement things similar to armv8 security. And also, how would this then support being used/extended out of tree?

Reply:
> Do we envisage similar requirements for other architectures that would need special handling here?

I don't know. Anecdotally, this is the first time that the system DT maintainers ran into this; they had no idea this was a possiblility, and they're running on multiple architectures.

I think the answer is probably "yes", but I also think we should proceed incrementally and gather more data with case-by-case code like this until a pattern emerges. It's not like new architectures are appearing every day.

> And also, how would this then support being used/extended out of tree?

Right now, the system devicetree specification itself is where the meaning of the execution level flags is defined, and the spec authors chose to mandate that the spec itself should be updated to add new architecture-specific definitions for this field. (I do plan on upstreaming this Arm v8-M convention into the spec as a replacement for a couple of previous, failed attempts to handle this case that we reverted in the spec; see https://github.com/devicetree-org/lopper/commit/0da7cbd70bc2c3417e5e909bdd90eea2a0a0c912 .)

TL;DR "you won't extend this out of tree for now". Let's gather data from publicly documented architectures that want to support this, then when the pattern emerges, we can extend the specification as needed.


So for now, this is a "heavyweight" operation